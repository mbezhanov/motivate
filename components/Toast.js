import React, { Component } from 'react';
import { StyleSheet, Text, Animated } from 'react-native';
import Colors from '../constants/Colors';
import Sizes from '../constants/Sizes';
import PropTypes from 'prop-types';

const ANIMATION_DURATION = 600; // milliseconds;
const TOAST_DURATION = 2000; // milliseconds

export default class Toast extends Component {
  constructor(props) {
    super(props);
    this.state = {
      opacity: new Animated.Value(0),
    };
    this.fadeOutTimeout = null;
  }

  componentDidMount() {
    Animated.timing(this.state.opacity, {
      toValue: 1,
      duration: ANIMATION_DURATION,
    }).start(() => {
      this.fadeOutTimeout = setTimeout(() => {
        Animated.timing(this.state.opacity, {
          toValue: 0,
          duration: ANIMATION_DURATION,
        }).start(() => {
          this.props.onComplete();
        });
      }, TOAST_DURATION);
    });
  }

  componentWillUnmount() {
    clearTimeout(this.fadeOutTimeout);
  }

  render() {
    const backgroundColor = this.props.transparent ? 'rgba(0, 0, 0, 0.5)' : Colors.BORDER;

    return (
      <Animated.View style={{...styles.container, opacity: this.state.opacity, backgroundColor }}>
        <Text style={styles.text}>{this.props.content}</Text>
      </Animated.View>
    );
  }
}

Toast.propTypes = {
  content: PropTypes.string.isRequired,
  transparent: PropTypes.bool,
  onComplete: PropTypes.func.isRequired,
};

const styles = StyleSheet.create({
  container: {
    width: '40%',
    borderRadius: Sizes.M,
    padding: Sizes.M / 2,
    margin: Sizes.M,
    flex: 0,
    alignItems: 'center',
    justifyContent: 'center',
    alignSelf: 'center',
  },
  text: {
    color: Colors.TEXT,
    fontSize: Sizes.M
  }
});
