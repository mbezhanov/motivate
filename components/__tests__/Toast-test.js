import React from 'react';
import renderer from 'react-test-renderer';
import snapshotDiff from 'snapshot-diff';
import Toast from '../Toast';

jest.mock('react-native/Libraries/Animated/src/Animated', () => {
  const ActualAnimated = jest.requireActual('react-native/Libraries/Animated/src/Animated');
  return {
    ...ActualAnimated,
    timing: (value, config) => {
      return {
        start: (callback) => {
          value.setValue(config.toValue);
          callback && callback();
        },
      };
    },
  };
});

describe('Toast', () => {
  let onComplete;
  let tree;

  beforeEach(() => {
    jest.useFakeTimers();
    onComplete = jest.fn();
    tree = renderer.create(<Toast content="foo" onComplete={onComplete} />);
  });

  it('renders correctly with opaque background', () => {
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders correctly with transparent background', () => {
    const a = tree.toJSON();
    const b = renderer.create(<Toast content="foo" onComplete={onComplete} transparent={true} />).toJSON();
    const diff = snapshotDiff(a, b);
    expect(diff).toMatchSnapshot();
  });

  it('fades in gradually and then fades out', () => {
    // test fade-in:
    expect(tree.getInstance().fadeOutTimeout).toBeDefined();
    expect(setTimeout).toHaveBeenCalledTimes(1);
    expect(onComplete.mock.calls.length).toBe(0);
    expect(tree.getInstance().state.opacity._value).toBe(1);

    // test fade-out:
    jest.runAllTimers();
    expect(onComplete.mock.calls.length).toBe(1);
    expect(tree.getInstance().state.opacity._value).toBe(0);
  });

  it('frees up allocated memory before unmounting', () => {
    tree.getInstance().componentWillUnmount();
    expect(clearTimeout).toHaveBeenCalledTimes(1);
    expect(clearTimeout).toHaveBeenCalledWith(tree.getInstance().fadeOutTimeout);
  });
});
