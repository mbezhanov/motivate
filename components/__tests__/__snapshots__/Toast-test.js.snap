// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast renders correctly with opaque background 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "center",
      "backgroundColor": "#555",
      "borderRadius": 16,
      "flex": 0,
      "justifyContent": "center",
      "margin": 16,
      "opacity": 1,
      "padding": 8,
      "width": "40%",
    }
  }
>
  <Text
    style={
      Object {
        "color": "#bbb",
        "fontSize": 16,
      }
    }
  >
    foo
  </Text>
</View>
`;

exports[`Toast renders correctly with transparent background 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,11 +1,11 @@
  <View
    style={
      Object {
        \\"alignItems\\": \\"center\\",
        \\"alignSelf\\": \\"center\\",
-       \\"backgroundColor\\": \\"#555\\",
+       \\"backgroundColor\\": \\"rgba(0, 0, 0, 0.5)\\",
        \\"borderRadius\\": 16,
        \\"flex\\": 0,
        \\"justifyContent\\": \\"center\\",
        \\"margin\\": 16,
        \\"opacity\\": 1,"
`;
